File: packages/core/lib/models/rider.dart
import 'package:equatable/equatable.dart';

class Rider extends Equatable {
  final String id;
  final String name;
  final String phone;
  final String? profilePhoto;
  final String vehicleType;
  final bool isOnline;
  final bool isVerified;
  final double rating;
  final int totalDeliveries;
  final double todayEarnings;
  final double latitude;
  final double longitude;

  const Rider({
    required this.id,
    required this.name,
    required this.phone,
    this.profilePhoto,
    required this.vehicleType,
    this.isOnline = false,
    this.isVerified = false,
    this.rating = 4.5,
    this.totalDeliveries = 0,
    this.todayEarnings = 0,
    this.latitude = 0,
    this.longitude = 0,
  });

  Rider copyWith({
    bool? isOnline,
    double? latitude,
    double? longitude,
    double? todayEarnings,
  }) =>
      Rider(
        id: id,
        name: name,
        phone: phone,
        profilePhoto: profilePhoto,
        vehicleType: vehicleType,
        isOnline: isOnline ?? this.isOnline,
        isVerified: isVerified,
        rating: rating,
        totalDeliveries: totalDeliveries,
        todayEarnings: todayEarnings ?? this.todayEarnings,
        latitude: latitude ?? this.latitude,
        longitude: longitude ?? this.longitude,
      );

  @override
  List<Object?> get props => [id, name, isOnline];
}

class DeliveryRequest extends Equatable {
  final String orderId;
  final String restaurantName;
  final String restaurantAddress;
  final String customerAddress;
  final double earnings;
  final double distanceKm;
  final int expiresInSeconds;

  const DeliveryRequest({
    required this.orderId,
    required this.restaurantName,
    required this.restaurantAddress,
    required this.customerAddress,
    required this.earnings,
    required this.distanceKm,
    this.expiresInSeconds = 15,
  });

  @override
  List<Object?> get props => [orderId];
}

File: backend/cmd/rider-service/internal/models/rider.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type Rider struct {
	ID                uuid.UUID  `json:"id" gorm:"type:uuid;primary_key;default:uuid_generate_v4()"`
	UserID            uuid.UUID  `json:"user_id" gorm:"not null;index"`
	VehicleType       string    `json:"vehicle_type" gorm:"not null"`
	VehicleNumber     string    `json:"vehicle_number" gorm:"not null"`
	VehicleModel      string    `json:"vehicle_model"`
	LicenseNumber     string    `json:"license_number"`
	CurrentLatitude   *float64  `json:"current_latitude"`
	CurrentLongitude  *float64  `json:"current_longitude"`
	LastLocationUpdate *time.Time `json:"last_location_update"`
	IsOnline          bool      `json:"is_online" gorm:"default:false"`
	IsAvailable       bool      `json:"is_available" gorm:"default:true"`
	Rating            float64   `json:"rating" gorm:"default:0.00"`
	TotalDeliveries   int       `json:"total_deliveries" gorm:"default:0"`
	TotalEarnings     float64   `json:"total_earnings" gorm:"default:0.00"`
	ZoneID            string    `json:"zone_id"`
	IsVerified        bool      `json:"is_verified" gorm:"default:false"`
	IsActive          bool      `json:"is_active" gorm:"default:true"`
	CreatedAt         time.Time `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt         time.Time `json:"updated_at" gorm:"autoUpdateTime"`
}

type RiderDocument struct {
	ID              uuid.UUID  `json:"id" gorm:"type:uuid;primary_key;default:uuid_generate_v4()"`
	RiderID         uuid.UUID  `json:"rider_id" gorm:"not null;index"`
	DocumentType    string     `json:"document_type" gorm:"not null"`
	DocumentURL     string     `json:"document_url" gorm:"not null"`
	ExpiryDate      *time.Time `json:"expiry_date"`
	Status          string     `json:"status" gorm:"default:'pending'"`
	RejectionReason string     `json:"rejection_reason"`
	VerifiedAt      *time.Time `json:"verified_at"`
	VerifiedBy      *uuid.UUID `json:"verified_by"`
	CreatedAt       time.Time  `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt       time.Time  `json:"updated_at" gorm:"autoUpdateTime"`
}

type RiderEarning struct {
	ID           uuid.UUID  `json:"id" gorm:"type:uuid;primary_key;default:uuid_generate_v4()"`
	RiderID      uuid.UUID  `json:"rider_id" gorm:"not null;index"`
	OrderID      uuid.UUID  `json:"order_id" gorm:"not null;index"`
	DeliveryFee  float64    `json:"delivery_fee" gorm:"not null"`
	TipAmount    float64    `json:"tip_amount" gorm:"default:0.00"`
	BonusAmount  float64    `json:"bonus_amount" gorm:"default:0.00"`
	TotalEarned  float64    `json:"total_earned" gorm:"not null"`
	EarnedAt     time.Time  `json:"earned_at" gorm:"autoCreateTime"`
}

type CreateRiderRequest struct {
	VehicleType   string `json:"vehicle_type" binding:"required"`
	VehicleNumber string `json:"vehicle_number" binding:"required"`
	VehicleModel  string `json:"vehicle_model"`
	LicenseNumber string `json:"license_number"`
	ZoneID        string `json:"zone_id"`
}

type UpdateRiderRequest struct {
	VehicleType       *string   `json:"vehicle_type"`
	VehicleNumber     *string   `json:"vehicle_number"`
	VehicleModel      *string   `json:"vehicle_model"`
	LicenseNumber     *string   `json:"license_number"`
	CurrentLatitude  *float64  `json:"current_latitude"`
	CurrentLongitude *float64  `json:"current_longitude"`
	IsOnline          *bool     `json:"is_online"`
	IsAvailable       *bool     `json:"is_available"`
	ZoneID            *string   `json:"zone_id"`
}

type UpdateLocationRequest struct {
	Latitude  float64 `json:"latitude" binding:"required"`
	Longitude float64 `json:"longitude" binding:"required"`
}

type ToggleOnlineRequest struct {
	IsOnline bool `json:"is_online" binding:"required"`
}

// BeforeCreate hook for Rider
func (r *Rider) BeforeCreate(tx *gorm.DB) error {
	if r.ID == uuid.Nil {
		r.ID = uuid.New()
	}
	return nil
}

// BeforeCreate hook for RiderDocument
func (rd *RiderDocument) BeforeCreate(tx *gorm.DB) error {
	if rd.ID == uuid.Nil {
		rd.ID = uuid.New()
	}
	return nil
}

// BeforeCreate hook for RiderEarning
func (re *RiderEarning) BeforeCreate(tx *gorm.DB) error {
	if re.ID == uuid.Nil {
		re.ID = uuid.New()
	}
	return nil
}
